### PR TITLE
[core] Support add partition for paimon

### PR DESCRIPTION
**Pull request by xuzifu666**
### Purpose
Currently paimon can not support add partition, due to add partition not need commit transaction，so make hms sync to achieve it.

Linked issue: close #xxx

### CODE REVIEW

File: paimon-core/src/main/java/org/apache/paimon/catalog/Catalog.java [hunk 1]
```diff
@@ -273,6 +273,17 @@ default void invalidateTable(Identifier identifier) {}
     void dropPartition(Identifier identifier, Map<String, String> partitions)
             throws TableNotExistException, PartitionNotExistException;
 
+    /**
+     * Create a partition for the specify table.
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/catalog/Catalog.java:277]: specified

File: paimon-hive/paimon-hive-catalog/src/main/java/org/apache/paimon/hive/HiveCatalog.java [hunk 2]
```diff
@@ -310,6 +310,27 @@ private Map<String, String> convertToProperties(Database database) {
         return properties;
     }
```
Comment (wwj6591812) [inline paimon-hive/paimon-hive-catalog/src/main/java/org/apache/paimon/hive/HiveCatalog.java:312]: @Override

File: paimon-core/src/main/java/org/apache/paimon/catalog/Catalog.java [hunk 3]
```diff
@@ -273,6 +273,17 @@ default void invalidateTable(Identifier identifier) {}
...
+     *
+     * @param identifier path of the table to add partition
+     * @param partitions the partition to be added
+     * @throws TableNotExistException if the table does not exist
+     * @throws PartitionNotExistException if the partition does not exist
+     */
+    void createPartition(Identifier identifier, Map<String, String> partitions)
+            throws TableNotExistException, PartitionNotExistException;
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/catalog/Catalog.java:285]: When throw PartitionNotExistException?

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkCatalog.java [hunk 4]
```diff
@@ -1145,8 +1148,28 @@ public final void createPartition(
...
+        try {
+            if (catalog instanceof CachingCatalog) {
+                catalog = ((CachingCatalog) catalog).wrapped();
+            }
+
+            if (catalog instanceof HiveCatalog) {
+                Identifier identifier = toIdentifier(tablePath);
+                catalog.createPartition(identifier, partitionSpec.getPartitionSpec());
+            } else {
+                throw new UnsupportedOperationException();
```
Comment (wwj6591812) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkCatalog.java:1166]: message?

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkCatalog.java [hunk 5]
```diff
@@ -1145,8 +1148,28 @@ public final void createPartition(
...
+            }
+        } catch (Catalog.TableNotExistException e) {
+            throw new CatalogException(e);
+        } catch (Catalog.PartitionNotExistException e) {
+            throw new RuntimeException(e);
```
Comment (wwj6591812) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkCatalog.java:1171]: Why throw RuntimeException?

File: paimon-hive/paimon-hive-catalog/src/main/java/org/apache/paimon/hive/HiveCatalog.java [hunk 6]
```diff
@@ -310,6 +310,27 @@ private Map<String, String> convertToProperties(Database database) {
...
 
+    public void createPartition(Identifier identifier, Map<String, String> partitionSpec)
+            throws TableNotExistException {
+        TableSchema tableSchema = getDataTableSchema(identifier);
+        if (!tableSchema.partitionKeys().isEmpty()
+                && new CoreOptions(tableSchema.options()).partitionedTableInMetastore()) {
+            try {
```
Comment (wwj6591812) [inline paimon-hive/paimon-hive-catalog/src/main/java/org/apache/paimon/hive/HiveCatalog.java:318]: Add a warn log if tableSchema.partitionKeys().isEmpty() or tableSchema.options()).partitionedTableInMetastore() is false?

**Comment by wwj6591812** [comment]
I'm not sure whether Paimon need this feature.
@JingsongLi  cc

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkCatalog.java [hunk 7]
```diff
@@ -1145,8 +1148,28 @@ public final void createPartition(
...
+            if (!ignoreIfExists) {
+                throw new PartitionAlreadyExistsException(getName(), tablePath, partitionSpec);
+            }
+        }
+        try {
+            if (catalog instanceof CachingCatalog) {
+                catalog = ((CachingCatalog) catalog).wrapped();
+            }
+
+            if (catalog instanceof HiveCatalog) {
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkCatalog.java:1162]: Why need to check HiveCatalog?

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkCatalog.java [hunk 8]
```diff
@@ -1145,8 +1148,28 @@ public final void createPartition(
...
-            throws CatalogException {
-        throw new UnsupportedOperationException();
+            throws CatalogException, PartitionAlreadyExistsException {
+        if (partitionExists(tablePath, partitionSpec)) {
+            if (!ignoreIfExists) {
+                throw new PartitionAlreadyExistsException(getName(), tablePath, partitionSpec);
+            }
+        }
+        try {
+            if (catalog instanceof CachingCatalog) {
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkCatalog.java:1158]: Why need to check CachingCatalog?

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkCatalog.java [hunk 9]
```diff
@@ -1145,8 +1148,28 @@ public final void createPartition(
...
+                catalog = ((CachingCatalog) catalog).wrapped();
+            }
+
+            if (catalog instanceof HiveCatalog) {
+                Identifier identifier = toIdentifier(tablePath);
+                catalog.createPartition(identifier, partitionSpec.getPartitionSpec());
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkCatalog.java:1164]: Just invoke this?

File: paimon-hive/paimon-hive-catalog/src/main/java/org/apache/paimon/hive/HiveCatalog.java [hunk 10]
```diff
@@ -310,6 +310,27 @@ private Map<String, String> convertToProperties(Database database) {
         return properties;
     }
 
+    public void createPartition(Identifier identifier, Map<String, String> partitionSpec)
+            throws TableNotExistException {
+        TableSchema tableSchema = getDataTableSchema(identifier);
+        if (!tableSchema.partitionKeys().isEmpty()
+                && new CoreOptions(tableSchema.options()).partitionedTableInMetastore()) {
```
Comment (JingsongLi) [inline paimon-hive/paimon-hive-catalog/src/main/java/org/apache/paimon/hive/HiveCatalog.java:317]: throw Unsupport exception if it is not partitioned table in metastore.

**Review by JingsongLi — CHANGES_REQUESTED** [review]
I think you should use `MetastoreClient` to add partition. The implementation can be in `AbstractCatalog`.

**Comment by xuzifu666** [comment]
> I think you should use `MetastoreClient` to add partition. The implementation can be in `AbstractCatalog`.

> I think you should use `MetastoreClient` to add partition. The implementation can be in `AbstractCatalog`.

Yes, move to AbstractCatalog maybe better, would change as it. Thanks @JingsongLi

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkCatalog.java (hunk 8 above)
Comment (xuzifu666) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkCatalog.java:1158]: image
Here catalog is CachingCatalog which wraped HiveCatalog, so handle it as CacheingCatalog firstly. @JingsongLi

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkCatalog.java (hunk 9 above)
Comment (xuzifu666) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkCatalog.java]: Had move the logic to AbstractCatalog, so Catalog API not contains createPartition API.

File: paimon-hive/paimon-hive-catalog/src/main/java/org/apache/paimon/hive/HiveCatalog.java (hunk 10 above)
Comment (xuzifu666) [reply paimon-hive/paimon-hive-catalog/src/main/java/org/apache/paimon/hive/HiveCatalog.java]: Done

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**